scipy/stats/_stats_py.py

```diff
@@ -10165,6 +10165,136 @@ def first_order(t):
+    _moment_result_object, n_samples=1, result_to_tuple=lambda x: (x,),
+    n_outputs=lambda kwds: _moment_outputs(kwds, [1, 2, 3, 4])
+)
+def lmoment(sample, order=None, *, axis=0, sorted=False, standardize=True):
+    r"""Compute L-moments of a sample from a continuous distribution
+
+    The L-moments of a probability distribution are summary statistics with
+    uses similar to those of conventional moments, but they are defined in
+    terms of the expected values of order statistics.
+    Sample L-moments are defined analogously to population L-moments, and
+    they can serve as estimators of population L-moments. They tend to be less
+    sensitive to extreme observations than conventional moments.
+
+    Parameters
+    ----------
+    sample : array_like
+        The real-valued sample whose L-moments are desired.
+    order : array_like, optional
+        The (positive integer) orders of the desired L-moments.
+        Must be a scalar or non-empty 1D array. Default is [1, 2, 3, 4].
+    axis : int, default=0
+        The axis along which to compute L-moments.
+    sorted : bool, default=False
+        Whether `sample` is already sorted in increasing order along `axis`.
+        If False (default), `sample` will be sorted.
+    standardize : bool, default=True
+        Whether to return L-moment ratios for orders 3 and higher.
+        L-moment ratios are analogous to standardized conventional
+        moments: they are the non-standardized L-moments divided
+        by the L-moment of order 2.
+
+    Returns
+    -------
+    lmoments : ndarray
+        The sample L-moments of order `order`.
+
+    See Also
+    --------
+    moment
+
+    References
+    ----------
+    .. [1] D. Bilkova. "L-Moments and TL-Moments as an Alternative Tool of
+           Statistical Data Analysis". Journal of Applied Mathematics and
+           Physics. 2014. :doi:`10.4236/jamp.2014.210104`
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from scipy import stats
+    >>> rng = np.random.default_rng(328458568356392)
+    >>> sample = rng.exponential(size=100000)
+    >>> stats.lmoment(sample)
+    array([1.00124272, 0.50111437, 0.3340092 , 0.16755338])
+
+    Note that the first four standardized population L-moments of the standard
+    exponential distribution are 1, 1/2, 1/3, and 1/6; the sample L-moments
+    provide reasonable estimates.
+
+    """
+    args = lmoment_iv(sample, order, axis, sorted, standardize)
+    sample, order, axis, sorted, standardize = args
+
+    n_moments = np.max(order)
+    k = np.arange(n_moments, dtype=sample.dtype)
+    prk = _prk(np.expand_dims(k, list(range(1, sample.ndim+1))), k)
+    bk = _br(sample, r=k)
+
+    n = sample.shape[-1]
+    bk[..., n:] = 0  # remove NaNs due to n_moments > n
+
+    lmoms = np.sum(prk * bk, axis=-1)
+    if standardize and n_moments > 2:
+        lmoms[2:] /= lmoms[1]
+
+    lmoms[n:] = np.nan  # add NaNs where appropriate
+    return lmoms[order-1]
```

Comment:
Well, we don't get that index error because it is caught by input validation:
```
ValueError: `order` must be a scalar or a non-empty array of positive integers.
```
But as I mentioned, I'll look into allowing `0`.

Reply:
After looking a little further, I see why I didn't include this in the first place.

Typical formulas break down for the zeroth l-moment in ways that they don't for the zeroth (regular) moment. For instance:

![image](https://github.com/user-attachments/assets/e9684c4c-4028-46e8-9f3f-a8b9e1c5f231)

The sum is over no elements, so we don't get to the fact that `binom(-1, j)` is NaN, but the denominator on the left is zero. I don't see a way of taking limits to get around the 0/0 because the sum is discrete. 

We run into similar problems with the formulas I use. We'd also need the -1st U-statistic.

Hoskings doesn't seem to define the 0th l-moment, or at least doesn't include it in the main discussion. It's clear from his definition that $r=0$ would be a special case.
![image](https://github.com/user-attachments/assets/016a9e68-a9e3-4ff5-b4cc-425e252a9798)

His `lmom` package can't calculate the 0th sample L-moment because it returns an integer number of consecutive sample L-moments beginning with 1. R's `EnvStats` function `lMoment` raises an error for `r=0`.

We currently document that the order must be (strictly) positive and raise an error when it is zero. Given the information above, this should not be considered incorrect behavior, and adding support for order 0 would be an enhancement.

I'm sure there are arguments for why the zeroth L-moment sould be defined as 1, but does a user need code to tell them that? Do you have a reference? Is that true when one of the observations is infinite or NaN? (If not, we 'd need to add special cases on top of the special case.)

I'd prefer to leave it out of this first PR, and if someone has a reference and interest, this could be extended in a follow-up PR.

Reply:
To my knowledge there aren't any authors who have used "zeroth" L-moment before.
But for the product-moments, this isn't the case; the 0th product moment is defined to be 1:

```pycon
>>> scipy.stats.moment([-1/12, 1/137, 3.14, 42], 0)
1.0
```

L-moments roughly describe the same properties that product moments describe, 
and all L-moments with orders $r < 2$ are equal to the product moments (i.e. the first L-moment == the first product-moment).
So if we simply continue this pattern, then the 0th L-moment would also be 1. Or at least, I can't think of anything else that it would be.

To understand a bit better, consider what the statistical (L- or product-) moments actually *do*:
They describe how the probability mass of a random variable is distributed.
And the way that (at least) the product-moments do this, is recursive (or inductive or something) by nature:

0. ???
1. The location of the probability mass w.r.t. **the total mass** -> The center of mass
2. The location of the probability mass w.r.t. **the first moment** -> The displacement
3. The location of the probability mass w.r.t. **the second moment** -> The "displacement of the displacement", i.e. the skewness
4. etc...

If we consider 0th moment to be the base case, then there's only one way we can define 0:

0. The `something` w.r.t. `something`

Now to figure out what `something` is, we consider the other moments and work our way backwards, and see that `something` is "the total mass".
So with that, we get:

0. The total mass w.r.t. to the total mass.

The total mass under a probability distribution is by definition 1, so the "w.r.t ..." part could even be left out. 
But if we leave it in, then the 0th moment is *also* 1 if some extra-terrestrial-quantum-alien decides to use 42 as the total probability mass.

👽